src/api.js

```diff
@@ -17,33 +17,433 @@ const getMessages = args => {
+      endDatetime: rawMessage.end_datetime ? new Date(rawMessage.end_datetime) : null,
+      pipeTarget:
+        rawMessage.options && rawMessage.options.pipe_target
+          ? rawMessage.options.pipe_target.map(parseMessage)
+          : null,
+      groupId:
+        rawMessage.options && rawMessage.options.group_info
+          ? rawMessage.options.group_info.group_id
+          : null,
+      compositionId: rawMessage.options ? rawMessage.options.composition_id : null,
+      queueName: rawMessage.queue_name
+    };
+    if (parsedMessage.enqueuedDatetime) {
+      if (parsedMessage.startedDatetime) {
+        parsedMessage.waitTime = parsedMessage.startedDatetime - parsedMessage.enqueuedDatetime;
+      } else {
+        parsedMessage.waitTime = loadDateTime - parsedMessage.enqueuedDatetime;
+      }
+    }
+    if (parsedMessage.startedDatetime) {
+      if (parsedMessage.endDatetime) {
+        parsedMessage.executionTime = parsedMessage.endDatetime - parsedMessage.startedDatetime;
+      } else {
+        parsedMessage.executionTime = loadDateTime - parsedMessage.startedDatetime;
+      }
+    }
+    if (parsedMessage.startedDatetime && parsedMessage.progress > 0 && !parsedMessage.endDatetime) {
+      parsedMessage.remainingTime =
+        ((loadDateTime - parsedMessage.startedDatetime) * (1 - parsedMessage.progress)) /
+        parsedMessage.progress;
+    }
+    return parsedMessage;
   };
-};
+
+  const messages = data.map(parseMessage);
+
+  /**
+   * Find the index of the message or composition in the array that has the target id
+   * @param target_id
+   * @param array
+   * @returns {*}
+   */
+  function findTargetIndex(target_id, array) {
+    return array.findIndex(element => element.messageId === target_id);
+  }
+
+  /**
+   * Find the index of an element that is before the element with the given index in a pipeline
+   * @param index
+   * @param array
+   * @returns {*}
+   */
+  function findPreviousElement(index, array) {
+    return array.findIndex(
+      msg =>
+        msg.pipeTarget && msg.pipeTarget.map(el => el.messageId).includes(array[index].messageId)
+    );
+  }
+  //grouping messages by composition_id
+  const compositions = {};
+  let i = 0;
+  while (i < messages.length) {
+    const compositionId = messages[i].compositionId;
+    if (compositionId) {
+      if (!compositions[compositionId]) {
+        compositions[compositionId] = [];
+      }
+      compositions[compositionId].push(messages.splice(i, 1)[0]);
+    } else {
+      i += 1;
+    }
+  }
+
+  /**
+   * Recursively adds to the composition all the messages from the pipe target that have not been enqueued
+   * @param composition_id
+   * @param message
+   */
+  function fillPipe(message) {
+    if (message.pipeTarget) {
+      message.pipeTarget.forEach(pipe_element => {
+        if (
+          compositions[message.compositionId].findIndex(
+            composition_element => composition_element.messageId === pipe_element.messageId
```

Comment:
you can also use destructuring
```suggestion
            ({messageId}) => messageId === targetMessage.messageId
```